core/names.js

```diff
@@ -62,19 +70,63 @@ Blockly.Names = function(reservedWords, opt_variablePrefix) {
 Blockly.Names.prototype.reset = function() {
   this.db_ = Object.create(null);
   this.dbReverse_ = Object.create(null);
+  this.variableMap_ = null;
+};
+
+/**
+ * Set the variable map that maps from variable name to variable object.
+ * @param {!Blockly.VariableMap} map The map to track.
+ * @package
+ */
+Blockly.Names.prototype.setVariableMap = function(map) {
```

Comment:
Should this just be passed in to the constructor of Names?

  At the same time I suggest that, I'm also not sure the Names object should depend on the workspace.